addons/l10n_pt_stock/models/stock_picking.py

```diff
@@ -0,0 +1,218 @@
+
+    ####################################
+    # GENERAL
+    ####################################
+
+    def _get_l10n_pt_stock_document_number(self):
+        self.ensure_one()
+        return f"stock_picking {self.name}"
+
+    def _get_l10n_pt_stock_gross_total(self):
+        return 0
+
+    ####################################
+    # HASH & ATCUD
+    ####################################
+    def _get_integrity_hash_fields(self):
+        if self.company_id.account_fiscal_country_id.code != 'PT':
+            return []
+        return ['date_done', 'l10n_pt_hashed_on', 'name']
+
+    @api.model
+    def _find_last_picking(self, company_id, picking_type):
+        return self.sudo().search([
+            ('company_id', '=', company_id),
+            ('picking_type_id', '=', picking_type.id),
+            ('l10n_pt_stock_inalterable_hash', '!=', False),
+        ], order='date_done desc', limit=1)
+
+    def _calculate_hashes(self, previous_hash=None):
+        if self.company_id.account_fiscal_country_id.code != 'PT':
+            return {}
+        self.l10n_pt_hashed_on = fields.Datetime.now()
+        docs_to_sign = [{
+            'id': picking.id,
+            'date': picking.date_done.strftime('%Y-%m-%d'),
+            'sorting_key': picking.date_done.isoformat(),
+            'system_entry_date': picking.l10n_pt_hashed_on.isoformat(timespec='seconds'),
+            'name': picking._get_l10n_pt_stock_document_number(),
+            'gross_total': float_repr(picking._get_l10n_pt_stock_gross_total(), precision_digits=2),
+            'previous_signature': previous_hash,
+        } for picking in self]
+        try:
+            return pt_hash_utils.sign_records(self.env, docs_to_sign)
+        except UserError as e:
+            self._message_log_batch(bodies={p.id: e.args[0] for p in self})
+            return {}
+
+    @api.depends('l10n_pt_stock_inalterable_hash')
+    def _compute_l10n_pt_stock_inalterable_hash_info(self):
+        for picking in self:
+            if picking.l10n_pt_stock_inalterable_hash:
+                hash_version, hash_str = picking.l10n_pt_stock_inalterable_hash.split("$")[1:]
+                picking.l10n_pt_stock_inalterable_hash_version = int(hash_version)
+                picking.l10n_pt_stock_inalterable_hash_short = hash_str[0] + hash_str[10] + hash_str[20] + hash_str[30]
+            else:
+                picking.l10n_pt_stock_inalterable_hash_version = False
+                picking.l10n_pt_stock_inalterable_hash_short = False
+
+    def _l10n_pt_stock_compute_missing_hashes(self, company_id=None):
+        """
+        Compute the hash/atcud for all records that do not have one yet
+        (because they were not printed/previewed yet)
+        """
+        company_id = company_id or self.company_id.id
+        picking_types = self.env['stock.picking.type'].search([
+            ('company_id', '=', company_id),
+            ('code', '=', 'outgoing'),
```

Comment:
I don't know if the QR code should be generated for dropshipping transfers as well. 
see
https://github.com/odoo/odoo/blob/b6f5832ecee324a3563f9d885db6b7af3db8bfde/addons/stock_dropshipping/models/stock.py#L57